### PR TITLE
Improvements to SettableListenableFuture and the Profiler

### PR DESCRIPTION
**Pull request by jentfoo**
I included both of these in the same commit/pull request since neither is _too_ big.
* The first is a new optional behavior in the SettableListenableFuture.  You can construct it so that if it has already completed it will just ignore future setResult or setFuture requests.  I know in the new ambush project we have to cancel these possible annoyingly (I think litesockets has a few places it is catching this IllegalStateException as well).
* The second improvement is to be able to start the profiler with an automatic stop time.  This is primarly so that it can make it safer to exposue the profiler in production systems.  By forcing a limited run time (as well as the fact that it wont run concurrently), it can become a much safer way to expose the profiler.

### CODE REVIEW

File: src/main/java/org/threadly/util/debug/Profiler.java [hunk 1]
```diff
@@ -188,7 +194,81 @@ public void start() {
...
+   */
+  private void start(Executor executor, final long sampleDurationInMillis, 
+                     SettableListenableFuture<String> completionFuture) {
     synchronized (startStopLock) {
+      if (sampleDurationInMillis > 0) {
+        // stop in case it's running, this allows us to simplify our start logic
+        stop();
+      }
+      if (completionFuture != null) {
+        stopFutures.add(completionFuture);
```
Comment (jentfoo) [inline src/main/java/org/threadly/util/debug/Profiler.java]: We create a list of futures that need to be provided a result when stop is called.  The only way for multiple futures can be on this list is if someone calls start with a 0 or lower sampleDurrationInMillis.  But this way ensures that if that does happen all futures will be completed (once the profile is stopped).  There is no way to return a future that will never complete (unless the profiler never stops).

File: src/main/java/org/threadly/util/debug/Profiler.java [hunk 2]
```diff
@@ -231,6 +311,30 @@ public void run() {
             throw ExceptionUtils.makeRuntime(e.getCause());
           }
         }
+        // start or schedule to handle run time limit
+        if (sampleDurationInMillis > 0) {
+          if (executor instanceof SimpleSchedulerInterface) {
```
Comment (jentfoo) [inline src/main/java/org/threadly/util/debug/Profiler.java]: The whole providing an executor is awkward...here if it is a scheduler (yucky ugly instanceof check I know) we schedule a task to call stop after the duration.  If an executor is null, or just a simple executor we create a new thread.  I did not think it would be worth taking a second thread on the executor to just block in a Thread.sleep() till we are ready to stop.

I could not let the profiling thread stop itself because it can't hold any references to this parent class which needs to have stop called.  Otherwise this class will never have the possibility of being garbage collected.  As implemented right now if you passed this in WITHOUT a timeout and then let the profiler reference go null, the profiler would automatically stop itself.  If you passed it in WITH a timeout, the profiler would continue to exist until this timeout thread ran, stopped it, then allowed for GC.

Let me know if you think we can make this cleaner in any way.  Honestly since these threads are pretty long running I wish we never accepted an Executor, and instead always just created threads.

**Comment by jentfoo** [comment]
FYI, the unit test failure on cloudbees is not accurate.  The issue is I switched the build system on 4.0.0-SNAPSHOT to use jacoco, and in doing that, the master build is not able to build PR's for the 4.X branch.

If I make any more PR's I will probably just move 4.X into master at this point, so it wont be an issue for future PR's but this one will just be marked as a failure on cloudbees.

File: src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java [hunk 3]
```diff
@@ -122,6 +155,11 @@ public void setResult(T result) {
...
    * provided in the {@link ExecutionException} on calls to {@link #get()}.
    * 
+   * If future has already completed and constructed with {@link #SettableListenableFuture()} or 
+   * {@code true} provided to {@link #SettableListenableFuture(boolean)} this will throw an 
+   * {@link IllegalStateException}.  If complete but constructed with a {@code false} this failure 
+   * result will be ignored.
+   * 
    * @param failure Throwable that caused failure during computation.
    */
   public void setFailure(Throwable failure) {
```
Comment (lwahlmeier) [inline src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java]: should probably change the overloaded variable name failure here.

Comment (jentfoo) [inline src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java]: not sure what you mean

File: src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java [hunk 4]
```diff
@@ -215,15 +255,20 @@ public void clearResult() {
   }
   
   // should be synchronized on resultLock before calling
-  private void setDone() {
+  private boolean setDone() {
```
Comment (lwahlmeier) [inline src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java]: not sure if we can get this in call cases, but this could probably be achieved through an atomicBoolean instead of needing a lock.

File: src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java [hunk 5]
```diff
@@ -131,7 +169,9 @@ public void setFailure(Throwable failure) {
     synchronized (resultLock) {
       this.failure = failure;
       
-      setDone();
+      if (! setDone()) {
```
Comment (lwahlmeier) [inline src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java]: Should probably not set the class variable until we setDone successfully

Comment (jentfoo) [inline src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java]: Ah yes, I will fix this.  Though it absolutely will make an atomic boolean impossible, since we don't want threads blocking on .get() to wake up before the result has been set (and notify happens in setDone() right now)

File: src/main/java/org/threadly/util/debug/Profiler.java [hunk 6]
```diff
@@ -188,7 +194,81 @@ public void start() {
...
+   * @param executor Executor or scheduler to use for execution as possible
+   * @param sampleDurationInMillis if greater than {@code 0} an automatic stop will occur after that many milliseconds 
+   * @param completionFuture If not {@code null} future will be called once the next {@link #stop()} is invoked
+   */
+  private void start(Executor executor, final long sampleDurationInMillis, 
+                     SettableListenableFuture<String> completionFuture) {
     synchronized (startStopLock) {
+      if (sampleDurationInMillis > 0) {
+        // stop in case it's running, this allows us to simplify our start logic
+        stop();
```
Comment (lwahlmeier) [inline src/main/java/org/threadly/util/debug/Profiler.java]: I would argue that we might want to throw here.  If someone starts 2 times, i think they are just doing something bad in code and we should just not allow it.  So we will complete the first one instead of the last one.

Comment (jentfoo) [inline src/main/java/org/threadly/util/debug/Profiler.java]: Yea, I considered that.  My thought was that someone calling this should feel confident that the Profiler will stop in the period of time provided.  If a simple .start() was called, and then we throw an exception here, we would never stop from the first stop.

Because we don't reset the stats on a stop, this has very little impact in the usage IMO (futures from other timed requests may complete early).

File: src/main/java/org/threadly/util/debug/Profiler.java (hunk 2 above)
Comment (lwahlmeier) [inline src/main/java/org/threadly/util/debug/Profiler.java]: I think I agree its probably just not worth taking in an executor and just create the extra thread.  If someone is running the profiler they already know there is extra over head.

Comment (jentfoo) [inline src/main/java/org/threadly/util/debug/Profiler.java]: We would need to release a 3.9.0 version if we want to do that in 4.0.0 release.  You think we should keep the complexity for now, or should we release a 3.9.0 to add the deprecation notices?

**Comment by jentfoo** [comment]
I just force pushed based off your recommendations, if it looks good go ahead and merge it.

**Comment by jentfoo** [comment]
Oh, but still let me know on if you think we should just stick with the complexity of the Executor in the profiler, or if we should release a 3.9.0 to mark it as deprecated.

File: src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java (hunk 3 above)
Comment (lwahlmeier) [inline src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java]: this.failure vs local failure

Comment (jentfoo) [inline src/main/java/org/threadly/concurrent/future/SettableListenableFuture.java]: it sets lower down, I think this is fine.

File: src/main/java/org/threadly/util/debug/Profiler.java (hunk 2 above)
Comment (lwahlmeier) [inline src/main/java/org/threadly/util/debug/Profiler.java]: hum, guess we can leave it for now and maybe do it fro 4.1.0

Comment (jentfoo) [inline src/main/java/org/threadly/util/debug/Profiler.java]: We would have to do it in a 5.X since it would be a breaking API.  Here is my thoughts:

* If I were to re-do it now, I probably would not provide the executor option

* Though is adds complexity, as long as we are happy with the implementation, let's leave it for now.  The part I am unsure on is it is clear how we can possibly schedule on it, or the fact that it continually uses a thread from the executor, etc, etc.

So basically if we think the implementation is clean, let's keep it.  If it's kind of awkward or hard to understand, then it's probably worth a 3.9.0 release to remove it.

**Comment by jentfoo** [comment]
I am going to go ahead and merge this.  I am going to open an issue to talk about if we should remove the executor option in the Profiler.  That change can come in another PR.
